File: week6-Python/sentimental-credit/credit.py
import sys


def main():
    number = get_input()
    if luhnsAlgorithm(number):
        output(number)
    else:
        print("INVALID")


def get_input():
    while True:
        try:
            n = int(input("Number: "))
            if n > 0:
                return n
        except ValueError:
            print("Not an integer")


def luhnsAlgorithm(number):
    sum1 = num = digits = sum2 = 0
    n = number
    while n > 0:
        num = ((n // 10) % 10) * 2
        while num > 0:
            digits = num % 10
            sum1 += digits
            num //= 10
        n //= 100

    digits = 0
    n = number
    while n > 0:
        digits = n % 10
        sum2 += digits
        n //= 100

    if ((sum1 + sum2) % 10) == 0:
        return True
    else:
        return False


def output(number):
    n = number // 10000000000000
    if n == 34 or n == 37:
        print("AMEX")

    n = number // 100000000000000
    if n > 50 and n < 56:
        print("MASTERCARD")

    n = number // (10 ** 12)
    if n == 4 or (n // (10 ** 3)) == 4:
        print("VISA")

    if n not in [34, 37] and not (50 < n < 56) and not (n == 4 and (n // 1000) == 4):
        print("INVALID")


main()

File: week1-C/credit/credit.c
#include <stdio.h>
#include <cs50.h>

long Number;
bool LuhnsAlgorithm();
bool Validatio;
void Output();

int main(void)
{
    Number = get_long("Number: ");
    Validatio = LuhnsAlgorithm();
    if (Validatio)
    {
        Output();
    }
    else
    {
        printf("INVALID\n");
    }
}

bool LuhnsAlgorithm()
{
    //algorithm that determine if a credit card number is valid
    int sum1 = 0, num = 0, digits = 0, sum2 = 0;
    long number = Number;
    // every other digit (from end) * 2
    while (number > 0)
    {
        num = ((number / 10) % 10) * 2;
        while (num > 0)
        {
            digits = num % 10;
            sum1 += digits;
            num /= 10;
        }
        number /= 100;
    }

    digits = 0;
    number = Number;
    // sum of digits that weren’t multiplied by 2
    while (number > 0)
    {
        digits = number % 10;
        sum2 += digits;
        number /= 100;
    }
    Validatio = ((sum1 + sum2) % 10) == 0 ? true : false;
    return Validatio;
}

void Output()
{
    /* type of card:
    AMEX\n - 15 STRT 34 OR 37
    MASTERCARD\n - 16 STRT 51, 52, 53, 54, 55
    VISA\n - 13 OR 16 STRT 4
    INVALID\n */
    long number;
    bool inval = true;
    number = Number / 10000000000000;
    if (number == 34 || number == 37)
    {
        printf("AMEX\n");
        inval = false;
    }
    number = Number / 100000000000000;
    if (number > 50 && number < 56)
    {
        printf("MASTERCARD\n");
        inval = false;
    }
    number = Number / 1000000000000;
    if (number == 4 || number / 1000 == 4)
    {
        printf("VISA\n");
        inval = false;
    }
    if (inval)
    {
        printf("INVALID\n");
    }
}